File: frontend/src/lib/mocks/learning-plan.ts
import type { LearningPlan } from "$lib/api/schemas";

// 学習プランのモック（§5.4 の例）。段階 1: チーム内資産（最優先）→ 段階 2: 外部資源候補。
// 実データ取得・Vector Search・進捗永続化は後続 issue。ここは UI レイアウト確認用。
export const mockLearningPlan: LearningPlan = {
  id: "plan-001",
  gap_concepts: ["distributed_caching", "ADR-0012", "RedisClient"],
  estimated_total_minutes: 70,
  steps: [
    {
      order: 1,
      completed: true,
      resource: {
        id: "r1",
        origin: "team",
        kind: "adr",
        title: "ADR-0012 DB 死活独立性ポリシー",
        source_ref: "ADR-0012",
        url: null,
        estimated_minutes: 10,
        priority: "required",
        dormant_days: 540, // 18 か月読まれていない（死蔵）
      },
    },
    {
      order: 2,
      completed: true,
      resource: {
        id: "r2",
        origin: "team",
        kind: "video",
        title: "勉強会「分散キャッシュ設計」",
        source_ref: "@alice 2023-Q4",
        url: null,
        estimated_minutes: 25,
        priority: "required",
        dormant_days: 210,
      },
    },
    {
      order: 3,
      completed: false,
      resource: {
        id: "r3",
        origin: "team",
        kind: "pr_comment",
        title: "PR #4523 レビュー議論",
        source_ref: "PR #4523 by @alice",
        url: null,
        estimated_minutes: 5,
        priority: "recommended",
        dormant_days: 95,
      },
    },
    {
      order: 4,
      completed: false,
      resource: {
        id: "r4",
        origin: "external",
        kind: "docs",
        title: "Redis 公式 Caching Patterns",
        source_ref: null,
        url: "https://redis.io/docs/latest/develop/use/patterns/",
        estimated_minutes: 20,
        priority: "supplementary",
      },
    },
    {
      order: 5,
      completed: false,
      resource: {
        id: "r5",
        origin: "external",
        kind: "book",
        title: '"Designing Data-Intensive Applications" Ch.7',
        source_ref: "Kleppmann",
        url: null,
        estimated_minutes: null,
        priority: "supplementary",
      },
    },
    {
      order: 6,
      completed: false,
      resource: {
        id: "r6",
        origin: "team",
        kind: "code",
        title: "RedisClient.ts を読む",
        source_ref: "src/lib/redis-client.ts",
        url: null,
        estimated_minutes: 10,
        priority: "hands_on",
      },
    },
  ],
};

File: backend/shared/shared/models/learning_plan.py
"""Shared learning-plan ORM (issue 035): ``LearningPlan`` / ``LearningStep`` / ``LearningResource``.

``learningPlanSchema``（``frontend/src/lib/api/schemas.ts``）に対応。生成パイプラインが資源とステップを埋める。
チーム資産（``origin="team"``）を外部資源より上段に並べるのは配信/生成側の責務。``project_id`` は索引のみ
（FK 無し、Job.project_id 流儀）。``plan_id``/``resource_id`` は shared 内で解決でき FK を張る。
"""

import uuid
from datetime import UTC, datetime

from sqlalchemy import JSON, Column, DateTime, UniqueConstraint, func
from sqlmodel import Field, SQLModel


class LearningResource(SQLModel, table=True):
    """A learning resource (team asset or external) referenced by a step."""

    __tablename__ = "learning_resources"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    project_id: uuid.UUID = Field(index=True, nullable=False)
    origin: str = Field(nullable=False)  # team / external
    kind: str = Field(nullable=False)  # adr / video / pr_comment / wiki / docs / book / article / code
    title: str = Field(nullable=False)
    source_ref: str | None = Field(default=None)
    url: str | None = Field(default=None)
    estimated_minutes: int | None = Field(default=None)
    priority: str = Field(nullable=False)  # required / recommended / supplementary / hands_on
    dormant_days: int | None = Field(default=None)
    origin_meta: dict = Field(default_factory=dict, sa_column=Column(JSON, nullable=False))


class LearningPlan(SQLModel, table=True):
    """A learning plan header (gap concepts + total minutes)."""

    __tablename__ = "learning_plans"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    project_id: uuid.UUID = Field(index=True, nullable=False)
    gap_concepts: list = Field(default_factory=list, sa_column=Column(JSON, nullable=False))  # list[str]
    estimated_total_minutes: int = Field(default=0, nullable=False)
    quiz_session_id: uuid.UUID | None = Field(default=None)
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC),
        sa_column=Column(DateTime(timezone=True), server_default=func.now(), nullable=False),
    )


class LearningStep(SQLModel, table=True):
    """An ordered step in a plan, pointing at one resource."""

    __tablename__ = "learning_steps"
    __table_args__ = (UniqueConstraint("plan_id", "order", name="uq_learning_steps_plan_order"),)

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    plan_id: uuid.UUID = Field(foreign_key="learning_plans.id", index=True, nullable=False)
    order: int = Field(nullable=False)
    completed: bool = Field(default=False, nullable=False)
    completed_at: datetime | None = Field(default=None, sa_column=Column(DateTime(timezone=True), nullable=True))
    resource_id: uuid.UUID = Field(foreign_key="learning_resources.id", nullable=False)
